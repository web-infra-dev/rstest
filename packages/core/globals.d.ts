declare global {
  const test: typeof import('@rstest/core')['test'];
  const describe: typeof import('@rstest/core')['describe'];
  const it: typeof import('@rstest/core')['it'];
  const expect: typeof import('@rstest/core')['expect'];
}
export {};
