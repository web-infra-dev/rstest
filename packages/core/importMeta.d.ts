interface ImportMeta {
  readonly rstest?: typeof import('@rstest/core');
}
