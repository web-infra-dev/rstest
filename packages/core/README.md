<picture>
  <img alt="Rstest Banner" src="https://assets.rspack.dev/rstest/rstest-banner.png">
</picture>

# Rstest

Rstest is a testing framework powered by Rspack.

## ðŸ“– License

Rstest is licensed under the [MIT License](https://github.com/web-infra-dev/rstest/blob/main/LICENSE).
