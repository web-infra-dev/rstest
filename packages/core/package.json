{
  "name": "@rstest/core",
  "version": "0.0.0",
  "description": "The Rsbuild-based test tool.",
  "bugs": {
    "url": "https://github.com/web-infra-dev/rstest/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/web-infra-dev/rstest",
    "directory": "packages/core"
  },
  "license": "MIT",
  "type": "module",
  "main": "./dist/index.js",
  "types": "./dist-types/index.d.ts",
  "bin": {
    "rstest": "./bin/rstest.js"
  },
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "default": "./dist/index.js"
    },
    "./globals": {
      "types": "./globals.d.ts"
    }
  },
  "files": [
    "bin",
    "dist",
    "dist-types",
    "compiled",
    "types.d.ts"
  ],
  "scripts": {
    "build": "rslib build",
    "dev": "rslib build --watch",
    "test": "npx rstest run --globals"
  },
  "dependencies": {
    "@rsbuild/core": "^1.2.19",
    "birpc": "2.2.0",
    "chai": "^5.2.0",
    "std-env": "^3.8.1",
    "tinypool": "^1.0.2"
  },
  "devDependencies": {
    "@babel/code-frame": "^7.26.2",
    "@jridgewell/trace-mapping": "0.3.25",
    "@rslib/core": "0.5.5",
    "@rstest/tsconfig": "workspace:*",
    "@types/babel__code-frame": "^7.0.6",
    "@vitest/expect": "^3.0.9",
    "cac": "^6.7.14",
    "picocolors": "^1.1.1",
    "rslog": "^1.2.3",
    "stacktrace-parser": "0.1.11",
    "tinyglobby": "^0.2.12"
  },
  "engines": {
    "node": ">=16.7.0"
  },
  "publishConfig": {
    "access": "public",
    "provenance": true,
    "registry": "https://registry.npmjs.org/"
  }
}
