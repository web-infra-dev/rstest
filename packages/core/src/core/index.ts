import type { RstestCommand, RstestConfig, RstestInstance } from '../types';
import { createContext } from './context';

export function createRstest(
  config: RstestConfig,
  command: RstestCommand,
  fileFilters: string[],
): RstestInstance {
  const context = createContext({ cwd: process.cwd(), command }, config);

  const runTests = async (): Promise<void> => {
    const { runTests } = await import('./runTests');
    await runTests(context, fileFilters);
  };

  return {
    runTests,
  };
}
