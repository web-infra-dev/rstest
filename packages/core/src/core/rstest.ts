import { existsSync } from 'node:fs';
import { SnapshotManager } from '@vitest/snapshot/manager';
import { join } from 'pathe';
import { isCI } from 'std-env';
import { withDefaultConfig } from '../config';
import { DefaultReporter } from '../reporter';
import { GithubActionsReporter } from '../reporter/githubActions';
import { JUnitReporter } from '../reporter/junit';
import { VerboseReporter } from '../reporter/verbose';
import type {
  NormalizedConfig,
  NormalizedProjectConfig,
  Project,
  ProjectContext,
  Reporter,
  RstestCommand,
  RstestConfig,
  RstestContext,
  Test,
  TestFileResult,
  TestResult,
} from '../types';
import { castArray, getAbsolutePath, TS_CONFIG_FILE } from '../utils';

/**
 * Only letters, numbers, "-", "_", and "$" are allowed.
 */
function formatEnvironmentName(name: string): string {
  return name.replace(/[^a-zA-Z0-9\-_$]/g, '_');
}

type Options = {
  cwd: string;
  command: RstestCommand;
  fileFilters?: string[];
  configFilePath?: string;
  projects: Project[];
};

export class Rstest implements RstestContext {
  public cwd: string;
  public command: RstestCommand;
  public fileFilters?: string[];
  public configFilePath?: string;
  public reporters: (Reporter | GithubActionsReporter | JUnitReporter)[];
  public snapshotManager: SnapshotManager;
  public version: string;
  public rootPath: string;
  public originalConfig: RstestConfig;
  public normalizedConfig: NormalizedConfig;
  public idMap: Map<string, Test> = new Map();
  public reporterResults: {
    results: TestFileResult[];
    testResults: TestResult[];
  } = {
    results: [],
    testResults: [],
  };
  public projects: ProjectContext[] = [];

  public constructor(
    {
      cwd = process.cwd(),
      command,
      fileFilters,
      configFilePath,
      projects,
    }: Options,
    userConfig: RstestConfig,
  ) {
    this.cwd = cwd;
    this.command = command;
    this.fileFilters = fileFilters;
    this.configFilePath = configFilePath;

    const rootPath = userConfig.root
      ? getAbsolutePath(cwd, userConfig.root)
      : cwd;

    const rstestConfig = withDefaultConfig(userConfig);
    const reporters =
      command !== 'list'
        ? createReporters(rstestConfig.reporters, {
            rootPath,
            config: rstestConfig,
          })
        : [];
    const snapshotManager = new SnapshotManager({
      updateSnapshot: rstestConfig.update ? 'all' : isCI ? 'none' : 'new',
    });
    this.reporters = reporters;
    this.snapshotManager = snapshotManager;
    this.version = RSTEST_VERSION;
    this.rootPath = rootPath;
    this.originalConfig = userConfig;
    this.normalizedConfig = rstestConfig;
    this.projects = projects.length
      ? projects.map((project) => {
          // TODO: support extend projects config
          const config = withDefaultConfig(
            project.config,
          ) as NormalizedProjectConfig;
          config.isolate = rstestConfig.isolate;
          config.source ??= {};

          if (!config.source.tsconfigPath) {
            const tsconfigPath = join(config.root, TS_CONFIG_FILE);

            if (existsSync(tsconfigPath)) {
              config.source.tsconfigPath = tsconfigPath;
            }
          }
          return {
            configFilePath: project.configFilePath,
            rootPath: config.root,
            name: config.name,
            environmentName: formatEnvironmentName(config.name),
            normalizedConfig: config,
          };
        })
      : [
          {
            configFilePath,
            rootPath,
            name: rstestConfig.name,
            environmentName: formatEnvironmentName(rstestConfig.name),
            normalizedConfig: rstestConfig,
          },
        ];
  }

  public updateReporterResultState(
    results: TestFileResult[],
    testResults: TestResult[],
    deletedEntries: string[] = [],
  ): void {
    // Update or add results
    results.forEach((item) => {
      const existingIndex = this.reporterResults.results.findIndex(
        (r) => r.testPath === item.testPath,
      );
      if (existingIndex !== -1) {
        this.reporterResults.results[existingIndex] = item;
      } else {
        this.reporterResults.results.push(item);
      }
    });

    // Clear existing test results for updated paths and add new ones
    const testPathsToUpdate = new Set(testResults.map((r) => r.testPath));
    this.reporterResults.testResults = this.reporterResults.testResults.filter(
      (r) => !testPathsToUpdate.has(r.testPath),
    );
    this.reporterResults.testResults.push(...testResults);

    // Remove deleted entries
    if (deletedEntries.length > 0) {
      const deletedPathsSet = new Set(deletedEntries);
      this.reporterResults.results = this.reporterResults.results.filter(
        (r) => !deletedPathsSet.has(r.testPath),
      );
      this.reporterResults.testResults =
        this.reporterResults.testResults.filter(
          (r) => !deletedPathsSet.has(r.testPath),
        );
    }
  }
}

const reportersMap: {
  default: typeof DefaultReporter;
  verbose: typeof VerboseReporter;
  'github-actions': typeof GithubActionsReporter;
  junit: typeof JUnitReporter;
} = {
  default: DefaultReporter,
  verbose: VerboseReporter,
  'github-actions': GithubActionsReporter,
  junit: JUnitReporter,
};

export type BuiltInReporterNames = keyof typeof reportersMap;

export function createReporters(
  reporters: RstestConfig['reporters'],
  initOptions: any = {},
): (Reporter | GithubActionsReporter | JUnitReporter)[] {
  const result = castArray(reporters).map((reporter) => {
    if (typeof reporter === 'string' || Array.isArray(reporter)) {
      const [name, options = {}] =
        typeof reporter === 'string' ? [reporter, {}] : reporter;
      // built-in reporters
      if (name in reportersMap) {
        const Reporter = reportersMap[name];
        return new Reporter({
          ...initOptions,
          options,
        });
      }

      // TODO: load third-party reporters
      throw new Error(
        `Reporter ${reporter} not found. Please install it or use a built-in reporter.`,
      );
    }

    return reporter;
  });

  return result;
}
