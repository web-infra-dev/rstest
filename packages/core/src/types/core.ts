import type { SnapshotManager } from '@vitest/snapshot/manager';
import type { NormalizedConfig, RstestConfig } from './config';
import type { Reporter } from './reporter';

export type RstestCommand = 'watch' | 'run';

export type RstestContext = {
  /** The Rstest core version. */
  version: string;
  /** The root path of current project. */
  rootPath: string;
  /** The original Rstest config passed from the createRstest method. */
  originalConfig: Readonly<RstestConfig>;
  /** The normalized Rstest config. */
  normalizedConfig: NormalizedConfig;
  /**
   * The command type.
   *
   * - dev: `rstest dev`
   * - run: `rstest run`
   */
  command: RstestCommand;
  reporters: Reporter[];
  snapshotManager: SnapshotManager;
};

export type RstestInstance = {
  runTests: () => Promise<void>;
};
