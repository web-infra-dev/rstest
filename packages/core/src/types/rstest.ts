import type { NormalizedConfig, RstestConfig } from './config';

export type RstestContext = {
  /** The Rstest core version. */
  version: string;
  /** The root path of current project. */
  rootPath: string;
  /** The original Rstest config passed from the createRstest method. */
  originalConfig: Readonly<RstestConfig>;
  /** The normalized Rstest config. */
  normalizedConfig: NormalizedConfig;
};

export type RstestInstance = {
  runTests: () => Promise<void>;
};
