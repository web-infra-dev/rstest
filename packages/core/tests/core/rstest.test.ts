import { join } from 'pathe';
import { Rstest } from '../../src/core/rstest';

process.env.DEBUG = 'false';

const rootPath = join(__dirname, '../..');

describe('rstest context', () => {
  it('should generate rstest context correctly', async () => {
    const rstestContext = new Rstest(
      {
        cwd: rootPath,
        command: 'run',
        projects: [],
      },
      {},
    );

    expect(rstestContext.projects[0]!.normalizedConfig).toMatchSnapshot();
  });

  it('should generate rstest context correctly with multiple projects', async () => {
    const rstestContext = new Rstest(
      {
        cwd: rootPath,
        command: 'run',
        projects: [
          {
            config: {
              root: join(rootPath, 'test-project'),
              name: 'test-project',
              include: ['<rootDir>/tests/**/*.test.ts'],
              setupFiles: '<rootDir>/scripts/rstest.setup.ts',
            },
          },
          {
            config: {
              root: 'test-project1',
              name: 'test-project1',
              setupFiles: ['<rootDir>/scripts/rstest.setup.ts'],
            },
          },
        ],
      },
      {},
    );

    expect(rstestContext.projects[0]!.normalizedConfig).toMatchSnapshot();
    expect(rstestContext.projects[1]!.normalizedConfig).toMatchSnapshot();
  });
});
