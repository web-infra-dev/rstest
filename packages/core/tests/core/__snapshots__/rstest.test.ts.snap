// Rstest Snapshot v1

exports[`rstest context > should generate rstest context correctly 1`] = `
{
  "clearMocks": false,
  "coverage": {
    "clean": true,
    "enabled": false,
    "exclude": [
      "**/node_modules/**",
      "**/[.]*",
      "**/dist/**",
      "**/test/**",
      "**/__tests__/**",
      "**/__mocks__/**",
      "**/*.d.ts",
      "**/*.{test,spec}.[jt]s",
      "**/*.{test,spec}.[c|m][jt]s",
      "**/*.{test,spec}.[jt]sx",
      "**/*.{test,spec}.[c|m][jt]sx",
    ],
    "provider": "istanbul",
    "reportOnFailure": false,
    "reporters": [
      "text",
      "html",
      "clover",
      "json",
    ],
    "reportsDirectory": "<ROOT>/packages/core/coverage",
  },
  "dev": undefined,
  "disableConsoleIntercept": false,
  "env": {},
  "exclude": {
    "override": false,
    "patterns": [
      "**/node_modules/**",
      "**/dist/**",
      "**/.{idea,git,cache,output,temp}/**",
      "**/dist/.rstest-temp",
    ],
  },
  "globals": false,
  "hideSkippedTests": false,
  "hookTimeout": 10000,
  "include": [
    "**/*.{test,spec}.?(c|m)[jt]s?(x)",
  ],
  "includeSource": [],
  "isolate": true,
  "maxConcurrency": 5,
  "name": "rstest",
  "output": undefined,
  "passWithNoTests": false,
  "pool": {
    "type": "forks",
  },
  "printConsoleTrace": false,
  "reporters": [
    "default",
  ],
  "resetMocks": false,
  "restoreMocks": false,
  "retry": 0,
  "root": "<ROOT>/packages/core",
  "setupFiles": [],
  "slowTestThreshold": 300,
  "snapshotFormat": {},
  "testEnvironment": "node",
  "testTimeout": 5000,
  "unstubEnvs": false,
  "unstubGlobals": false,
  "update": false,
}
`;

exports[`rstest context > should generate rstest context correctly with multiple projects 1`] = `
{
  "clearMocks": false,
  "coverage": {
    "clean": true,
    "enabled": false,
    "exclude": [
      "**/node_modules/**",
      "**/[.]*",
      "**/dist/**",
      "**/test/**",
      "**/__tests__/**",
      "**/__mocks__/**",
      "**/*.d.ts",
      "**/*.{test,spec}.[jt]s",
      "**/*.{test,spec}.[c|m][jt]s",
      "**/*.{test,spec}.[jt]sx",
      "**/*.{test,spec}.[c|m][jt]sx",
    ],
    "provider": "istanbul",
    "reportOnFailure": false,
    "reporters": [
      "text",
      "html",
      "clover",
      "json",
    ],
    "reportsDirectory": "<ROOT>/packages/core/coverage",
  },
  "dev": undefined,
  "disableConsoleIntercept": false,
  "env": {},
  "exclude": {
    "override": false,
    "patterns": [
      "**/node_modules/**",
      "**/dist/**",
      "**/.{idea,git,cache,output,temp}/**",
      "**/dist/.rstest-temp",
    ],
  },
  "globals": false,
  "hideSkippedTests": false,
  "hookTimeout": 10000,
  "include": [
    "<ROOT>/packages/core/test-project/tests/**/*.test.ts",
  ],
  "includeSource": [],
  "isolate": true,
  "maxConcurrency": 5,
  "name": "test-project",
  "output": undefined,
  "passWithNoTests": false,
  "pool": {
    "type": "forks",
  },
  "printConsoleTrace": false,
  "reporters": [
    "default",
  ],
  "resetMocks": false,
  "restoreMocks": false,
  "retry": 0,
  "root": "<ROOT>/packages/core/test-project",
  "setupFiles": [
    "<ROOT>/packages/core/test-project/scripts/rstest.setup.ts",
  ],
  "slowTestThreshold": 300,
  "snapshotFormat": {},
  "source": {},
  "testEnvironment": "node",
  "testTimeout": 5000,
  "unstubEnvs": false,
  "unstubGlobals": false,
  "update": false,
}
`;

exports[`rstest context > should generate rstest context correctly with multiple projects 2`] = `
{
  "clearMocks": false,
  "coverage": {
    "clean": true,
    "enabled": false,
    "exclude": [
      "**/node_modules/**",
      "**/[.]*",
      "**/dist/**",
      "**/test/**",
      "**/__tests__/**",
      "**/__mocks__/**",
      "**/*.d.ts",
      "**/*.{test,spec}.[jt]s",
      "**/*.{test,spec}.[c|m][jt]s",
      "**/*.{test,spec}.[jt]sx",
      "**/*.{test,spec}.[c|m][jt]sx",
    ],
    "provider": "istanbul",
    "reportOnFailure": false,
    "reporters": [
      "text",
      "html",
      "clover",
      "json",
    ],
    "reportsDirectory": "<ROOT>/packages/core/coverage",
  },
  "dev": undefined,
  "disableConsoleIntercept": false,
  "env": {},
  "exclude": {
    "override": false,
    "patterns": [
      "**/node_modules/**",
      "**/dist/**",
      "**/.{idea,git,cache,output,temp}/**",
      "**/dist/.rstest-temp",
    ],
  },
  "globals": false,
  "hideSkippedTests": false,
  "hookTimeout": 10000,
  "include": [
    "**/*.{test,spec}.?(c|m)[jt]s?(x)",
  ],
  "includeSource": [],
  "isolate": true,
  "maxConcurrency": 5,
  "name": "test-project1",
  "output": undefined,
  "passWithNoTests": false,
  "pool": {
    "type": "forks",
  },
  "printConsoleTrace": false,
  "reporters": [
    "default",
  ],
  "resetMocks": false,
  "restoreMocks": false,
  "retry": 0,
  "root": "<ROOT>/packages/core/test-project1",
  "setupFiles": [
    "<ROOT>/packages/core/test-project1/scripts/rstest.setup.ts",
  ],
  "slowTestThreshold": 300,
  "snapshotFormat": {},
  "source": {},
  "testEnvironment": "node",
  "testTimeout": 5000,
  "unstubEnvs": false,
  "unstubGlobals": false,
  "update": false,
}
`;
